src/sample.rs

```diff
@@ -0,0 +1,330 @@
+    // are slower than array index swapping.
+    if amount >= length / 2 {
+        sample_indices_inplace(rng, length, amount)
+    } else {
+        sample_indices_cache(rng, length, amount)
+    }
+}
+
+/// Sample an amount of indices using an inplace partial fisher yates method.
+///
+/// This allocates the entire `length` of indices and randomizes only the first `amount`.
+/// It then truncates to `amount` and returns.
+///
+/// This is better than using a HashMap "cache" when `amount >= length / 2` since it does not
+/// require allocating an extra cache and is much faster.
+fn sample_indices_inplace<R>(rng: &mut R, length: usize, amount: usize) -> Vec<usize>
+    where R: Rng,
+{
+    debug_assert!(amount <= length);
+    let amount = if amount == length {
+        // It isn't necessary to shuffle the final element if we are shuffling
+        // the whole array... it would just be shuffled with itself
+        amount - 1
+    } else {
+        amount
+    };
+
+    let mut indices: Vec<usize> = Vec::with_capacity(length);
+    indices.extend(0..length);
+    for i in 0..amount {
+        let j: usize = rng.gen_range(i, length);
+        let tmp = indices[i];
+        indices[i] = indices[j];
+        indices[j] = tmp;
+    }
+    indices.truncate(amount);
+    indices
+}
+
+
+/// This method performs a partial fisher-yates on a range of indices using a HashMap
+/// as a cache to record potential collisions.
+///
+/// The cache avoids allocating the entire `length` of values. This is especially useful when
+/// `amount <<< length`, i.e. select 3 non-repeating from 1_000_000
+fn sample_indices_cache<R>(
+    rng: &mut R,
+    length: usize,
+    amount: usize,
+) -> Vec<usize>
+    where R: Rng,
+{
+    debug_assert!(amount <= length);
+    let mut cache = HashMap::with_capacity(amount);
+    let mut out = Vec::with_capacity(amount);
+    for i in 0..amount {
+        let j: usize = rng.gen_range(i, length);
+
+        // equiv: let tmp = slice[i];
+        let tmp = match cache.get(&i) {
+            Some(e) => *e,
+            None => i,
+        };
+
+        // equiv: slice[i] = slice[j];
+        let x = match cache.get(&j) {
+            Some(x) => *x,
+            None => j,
+        };
+
+        // equiv: slice[j] = tmp;
+        cache.insert(j, tmp);
+
+        // note that in the inplace version, slice[i] is automatically "returned" value
+        out.push(x);
+    }
+    out
+}
+
+impl<'a, T: Clone> Sample for &'a [T] {
+    type Sampled = Vec<T>;
+
+    fn sample<R: Rng>(&self, rng: &mut R, amount: usize) -> Vec<T> {
+        let indices = sample_indices(rng, self.len(), amount);
+
+        let mut out = Vec::with_capacity(amount);
+        out.extend(indices.iter().map(|i| self[*i].clone()));
+        out
+    }
+}
+
+impl<'a, T: Clone> Sample for Vec<T> {
+    type Sampled = Vec<T>;
+
+    fn sample<R: Rng>(&self, rng: &mut R, amount: usize) -> Vec<T> {
+        self.as_slice().sample(rng, amount)
```

Comment:
So this just uses the slice representation directly? This makes me wonder whether the trait-based approach is worth it at all.